File: binding/include/opengl/libqopenglpaintdevice.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qopenglpaintdevice.html
pub const qopenglpaintdevice = struct {

    /// New constructs a new QOpenGLPaintDevice object.
    pub fn New() QtC.QOpenGLPaintDevice {
        return qtc.QOpenGLPaintDevice_new();
    }


    /// New2 constructs a new QOpenGLPaintDevice object.
    pub fn New2(size: ?*anyopaque) QtC.QOpenGLPaintDevice {
        return qtc.QOpenGLPaintDevice_new2(@ptrCast(size));
    }


    /// New3 constructs a new QOpenGLPaintDevice object.
    pub fn New3(width: i32, height: i32) QtC.QOpenGLPaintDevice {
        return qtc.QOpenGLPaintDevice_new3(width, height);
    }


    pub fn DevType(self: ?*anyopaque, ) i32 {
        return qtc.QOpenGLPaintDevice_DevType(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnDevType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QOpenGLPaintDevice_OnDevType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseDevType(self: ?*anyopaque, ) i32 {
        return qtc.QOpenGLPaintDevice_QBaseDevType(@ptrCast(self));
    }

    pub fn PaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QOpenGLPaintDevice_PaintEngine(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEngine(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QPaintEngine) void {
        qtc.QOpenGLPaintDevice_OnPaintEngine(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QOpenGLPaintDevice_QBasePaintEngine(@ptrCast(self));
    }

    pub fn Context(self: ?*anyopaque, ) QtC.QOpenGLContext {
        return qtc.QOpenGLPaintDevice_Context(@ptrCast(self));
    }

    pub fn Size(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QOpenGLPaintDevice_Size(@ptrCast(self));
    }

    pub fn SetSize(self: ?*anyopaque, size: ?*anyopaque) void {
        qtc.QOpenGLPaintDevice_SetSize(@ptrCast(self), @ptrCast(size));
    }

    pub fn SetDevicePixelRatio(self: ?*anyopaque, devicePixelRatio: f64) void {
        qtc.QOpenGLPaintDevice_SetDevicePixelRatio(@ptrCast(self), devicePixelRatio);
    }

    pub fn DotsPerMeterX(self: ?*anyopaque, ) f64 {
        return qtc.QOpenGLPaintDevice_DotsPerMeterX(@ptrCast(self));
    }

    pub fn DotsPerMeterY(self: ?*anyopaque, ) f64 {
        return qtc.QOpenGLPaintDevice_DotsPerMeterY(@ptrCast(self));
    }

    pub fn SetDotsPerMeterX(self: ?*anyopaque, dotsPerMeterX: f64) void {
        qtc.QOpenGLPaintDevice_SetDotsPerMeterX(@ptrCast(self), dotsPerMeterX);
    }

    pub fn SetDotsPerMeterY(self: ?*anyopaque, dotsPerMeterY: f64) void {
        qtc.QOpenGLPaintDevice_SetDotsPerMeterY(@ptrCast(self), dotsPerMeterY);
    }

    pub fn SetPaintFlipped(self: ?*anyopaque, flipped: bool) void {
        qtc.QOpenGLPaintDevice_SetPaintFlipped(@ptrCast(self), flipped);
    }

    pub fn PaintFlipped(self: ?*anyopaque, ) bool {
        return qtc.QOpenGLPaintDevice_PaintFlipped(@ptrCast(self));
    }

    pub fn EnsureActiveTarget(self: ?*anyopaque, ) void {
        qtc.QOpenGLPaintDevice_EnsureActiveTarget(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnEnsureActiveTarget(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QOpenGLPaintDevice_OnEnsureActiveTarget(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEnsureActiveTarget(self: ?*anyopaque, ) void {
        qtc.QOpenGLPaintDevice_QBaseEnsureActiveTarget(@ptrCast(self));
    }

    pub fn Metric(self: ?*anyopaque, metric: i32) i32 {
        return qtc.QOpenGLPaintDevice_Metric(@ptrCast(self), @intCast(metric));
    }

    /// Allows for overriding the related default method
    pub fn OnMetric(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) i32) void {
        qtc.QOpenGLPaintDevice_OnMetric(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMetric(self: ?*anyopaque, metric: i32) i32 {
        return qtc.QOpenGLPaintDevice_QBaseMetric(@ptrCast(self), @intCast(metric));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QOpenGLPaintDevice_Delete(@ptrCast(self));
    }
};

File: binding/src/opengl/libqopenglpaintdevice.cpp
#include <QOpenGLContext>
#include <QOpenGLPaintDevice>
#include <QPaintDevice>
#include <QPaintEngine>
#include <QSize>
#include <qopenglpaintdevice.h>
#include "libqopenglpaintdevice.h"
#include "libqopenglpaintdevice.hxx"

QOpenGLPaintDevice* QOpenGLPaintDevice_new() {
	 return new VirtualQOpenGLPaintDevice();
}

QOpenGLPaintDevice* QOpenGLPaintDevice_new2(const QSize* size) {
	 return new VirtualQOpenGLPaintDevice(*size);
}

QOpenGLPaintDevice* QOpenGLPaintDevice_new3(int width, int height) {
	 return new VirtualQOpenGLPaintDevice(width, height);
}

int QOpenGLPaintDevice_DevType(const QOpenGLPaintDevice* self) {
	return self->devType();
}

QPaintEngine* QOpenGLPaintDevice_PaintEngine(const QOpenGLPaintDevice* self) {
	return self->paintEngine();
}

QOpenGLContext* QOpenGLPaintDevice_Context(const QOpenGLPaintDevice* self) {
	return self->context();
}

QSize* QOpenGLPaintDevice_Size(const QOpenGLPaintDevice* self) {
	return new QSize(self->size());
}

void QOpenGLPaintDevice_SetSize(QOpenGLPaintDevice* self, const QSize* size) {
	self->setSize(*size);
}

void QOpenGLPaintDevice_SetDevicePixelRatio(QOpenGLPaintDevice* self, double devicePixelRatio) {
	self->setDevicePixelRatio(devicePixelRatio);
}

double QOpenGLPaintDevice_DotsPerMeterX(const QOpenGLPaintDevice* self) {
	return self->dotsPerMeterX();
}

double QOpenGLPaintDevice_DotsPerMeterY(const QOpenGLPaintDevice* self) {
	return self->dotsPerMeterY();
}

void QOpenGLPaintDevice_SetDotsPerMeterX(QOpenGLPaintDevice* self, double dotsPerMeterX) {
	self->setDotsPerMeterX(dotsPerMeterX);
}

void QOpenGLPaintDevice_SetDotsPerMeterY(QOpenGLPaintDevice* self, double dotsPerMeterY) {
	self->setDotsPerMeterY(dotsPerMeterY);
}

void QOpenGLPaintDevice_SetPaintFlipped(QOpenGLPaintDevice* self, bool flipped) {
	self->setPaintFlipped(flipped);
}

bool QOpenGLPaintDevice_PaintFlipped(const QOpenGLPaintDevice* self) {
	return self->paintFlipped();
}

void QOpenGLPaintDevice_EnsureActiveTarget(QOpenGLPaintDevice* self) {
	self->ensureActiveTarget();
}

// Base class handler implementation
int QOpenGLPaintDevice_QBaseDevType(const QOpenGLPaintDevice* self) {
	auto* vqopenglpaintdevice = dynamic_cast<const VirtualQOpenGLPaintDevice*>(self);
	if (vqopenglpaintdevice && vqopenglpaintdevice->isVirtualQOpenGLPaintDevice) {
vqopenglpaintdevice->setQOpenGLPaintDevice_DevType_IsBase(true);
	return vqopenglpaintdevice->devType();
}
}

// Auxiliary method to allow providing re-implementation
void QOpenGLPaintDevice_OnDevType(const QOpenGLPaintDevice* self, intptr_t slot) {
	auto* vqopenglpaintdevice = dynamic_cast<const VirtualQOpenGLPaintDevice*>(self);
	if (vqopenglpaintdevice && vqopenglpaintdevice->isVirtualQOpenGLPaintDevice) {
vqopenglpaintdevice->setQOpenGLPaintDevice_DevType_Callback(reinterpret_cast<VirtualQOpenGLPaintDevice::QOpenGLPaintDevice_DevType_Callback>(slot));
}
}

// Base class handler implementation
QPaintEngine* QOpenGLPaintDevice_QBasePaintEngine(const QOpenGLPaintDevice* self) {
	auto* vqopenglpaintdevice = dynamic_cast<const VirtualQOpenGLPaintDevice*>(self);
	if (vqopenglpaintdevice && vqopenglpaintdevice->isVirtualQOpenGLPaintDevice) {
vqopenglpaintdevice->setQOpenGLPaintDevice_PaintEngine_IsBase(true);
	return vqopenglpaintdevice->paintEngine();
}
}

// Auxiliary method to allow providing re-implementation
void QOpenGLPaintDevice_OnPaintEngine(const QOpenGLPaintDevice* self, intptr_t slot) {
	auto* vqopenglpaintdevice = dynamic_cast<const VirtualQOpenGLPaintDevice*>(self);
	if (vqopenglpaintdevice && vqopenglpaintdevice->isVirtualQOpenGLPaintDevice) {
vqopenglpaintdevice->setQOpenGLPaintDevice_PaintEngine_Callback(reinterpret_cast<VirtualQOpenGLPaintDevice::QOpenGLPaintDevice_PaintEngine_Callback>(slot));
}
}

// Base class handler implementation
void QOpenGLPaintDevice_QBaseEnsureActiveTarget(QOpenGLPaintDevice* self) {
	auto* vqopenglpaintdevice = dynamic_cast<VirtualQOpenGLPaintDevice*>(self);
	if (vqopenglpaintdevice && vqopenglpaintdevice->isVirtualQOpenGLPaintDevice) {
vqopenglpaintdevice->setQOpenGLPaintDevice_EnsureActiveTarget_IsBase(true);
	vqopenglpaintdevice->ensureActiveTarget();
}
}

// Auxiliary method to allow providing re-implementation
void QOpenGLPaintDevice_OnEnsureActiveTarget(QOpenGLPaintDevice* self, intptr_t slot) {
	auto* vqopenglpaintdevice = dynamic_cast<VirtualQOpenGLPaintDevice*>(self);
	if (vqopenglpaintdevice && vqopenglpaintdevice->isVirtualQOpenGLPaintDevice) {
vqopenglpaintdevice->setQOpenGLPaintDevice_EnsureActiveTarget_Callback(reinterpret_cast<VirtualQOpenGLPaintDevice::QOpenGLPaintDevice_EnsureActiveTarget_Callback>(slot));
}
}

void QOpenGLPaintDevice_Delete(QOpenGLPaintDevice* self) {
    delete self;
}

